File: frontend/src/types/group.ts
// SPDX-FileCopyrightText: 2025 Weibo, Inc.
//
// SPDX-License-Identifier: Apache-2.0

/**
 * Group and GroupMember TypeScript types
 */

export enum GroupRole {
  OWNER = 'Owner',
  MAINTAINER = 'Maintainer',
  DEVELOPER = 'Developer',
  REPORTER = 'Reporter',
}

export enum ResourceSource {
  PUBLIC = 'public',
  PERSONAL = 'personal',
  GROUP = 'group',
  SHARED = 'shared',
}

export interface GroupBase {
  name: string
  display_name?: string | null
  description?: string | null
}

export interface GroupCreate {
  name: string
  display_name?: string | null
  parent_path?: string | null
  description?: string | null
}

export interface GroupUpdate {
  display_name?: string | null
  description?: string | null
}

export interface GroupInDB {
  id: number
  name: string
  display_name?: string | null
  owner_user_id: number
  visibility: string
  description?: string | null
  is_active: boolean
  created_at: string
  updated_at: string
}

export interface GroupListItem {
  id: number
  name: string
  display_name?: string | null
  parent_path?: string | null
  simple_name?: string | null
  description?: string | null
  member_count: number
  resource_count: number
  my_role: GroupRole
  created_at: string
  path_depth?: number
}

export interface GroupListResponse {
  total: number
  items: GroupListItem[]
}

export interface GroupDetail extends GroupInDB {
  display_name?: string | null
  parent_name?: string | null
  member_count: number
  resource_count: number
  my_role: GroupRole
  owner_name: string
}

// GroupMember Types
export interface GroupMemberBase {
  user_id: number
  role: GroupRole
}

export interface GroupMemberInvite {
  user_name: string
  role?: GroupRole
}

export interface GroupMemberUpdate {
  role: GroupRole
}

export interface GroupMemberInDB extends GroupMemberBase {
  id: number
  group_name: string
  invited_by_user_id?: number | null
  is_active: boolean
  created_at: string
  updated_at: string
}

export interface GroupMemberListItem {
  id: number
  user_id: number
  user_name: string
  role: GroupRole
  invited_by_user_id?: number | null
  invited_by_user_name?: string | null
  created_at: string
}

export interface GroupMemberListResponse {
  total: number
  items: GroupMemberListItem[]
}

export interface TransferOwnershipRequest {
  new_owner_user_id: number
}

export interface InviteAllUsersRequest {
  role?: GroupRole
}

// Resource with source information
export interface ResourceWithSource {
  source: ResourceSource
  group_name?: string | null
}

File: backend/app/schemas/group.py
# SPDX-FileCopyrightText: 2025 Weibo, Inc.
#
# SPDX-License-Identifier: Apache-2.0

"""
Pydantic schemas for Group and GroupMember
"""
from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, Field, field_validator


class GroupBase(BaseModel):
    """Base schema for Group"""
    name: str = Field(..., min_length=1, max_length=100, description="Unique group name (immutable)")
    display_name: str = Field(..., min_length=1, max_length=100, description="Display name (can be modified)")
    description: Optional[str] = Field(None, description="Group description")
    visibility: str = Field(default="private", description="Group visibility: private, public")
    parent_id: Optional[int] = Field(None, description="Parent group ID for hierarchical organization")

    @field_validator("name")
    @classmethod
    def validate_name(cls, v: str) -> str:
        """Validate group name format"""
        if not v.replace("-", "").replace("_", "").isalnum():
            raise ValueError("Group name must contain only alphanumeric characters, hyphens, and underscores")
        return v


class GroupCreate(GroupBase):
    """Schema for creating a Group"""
    pass


class GroupUpdate(BaseModel):
    """Schema for updating a Group"""
    display_name: Optional[str] = Field(None, min_length=1, max_length=100)
    description: Optional[str] = None
    visibility: Optional[str] = None
    parent_id: Optional[int] = None


class GroupResponse(GroupBase):
    """Schema for Group response"""
    id: int
    owner_user_id: int
    is_active: bool
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class GroupWithMembers(GroupResponse):
    """Schema for Group with member count"""
    member_count: int = 0
    role: Optional[str] = None  # Current user's role in the group


class GroupMemberBase(BaseModel):
    """Base schema for GroupMember"""
    group_name: str = Field(..., description="Group name")
    user_id: int = Field(..., description="User ID")
    role: str = Field(default="member", description="Role: owner, admin, member")


class GroupMemberCreate(GroupMemberBase):
    """Schema for creating a GroupMember"""
    pass


class GroupMemberUpdate(BaseModel):
    """Schema for updating a GroupMember"""
    role: Optional[str] = Field(None, description="Update member role")


class GroupMemberResponse(GroupMemberBase):
    """Schema for GroupMember response"""
    id: int
    invited_by_user_id: Optional[int]
    is_active: bool
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class GroupMemberWithUser(GroupMemberResponse):
    """Schema for GroupMember with user details"""
    user_name: str
    user_email: Optional[str] = None
